daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetInputter.scala

```diff
@@ -21,79 +21,70 @@ import java.net.URI
- * transferring control to the contentHandler so it can load them.
- * - After it is done with the current event, it calls inputter.next to get the next event, and that
- * increments the currentIndex and cleans out the event at the previous index
- * - This process continues until the event at currentIndex either contains an EndDocument event or
- * the currentIndex is the last in the batch. If it is the former, the endDocumentReceived flag is
- * set to true and hasNext will return false. If it is the latter, the next batch of events will be
- * queued by transferring control to the contentHandler so it can load them.
- * - This ends the unparse process, and the unparseResult and/or any Errors are set on a single element
- * array containing response events. We call resumeFinal passing along that array, terminating this
- * thread and resuming the contentHandler for the last time.
- *
- * @param unparseContentHandler producer coroutine that sends the SAXInfosetEvent to this class
- * @param dp dataprocessor that we use to kickstart the unparse process and that consumes the
- *           currentEvent
- * @param output  outputChannel of choice where the unparsed data is stored
+ * @param unparseContentHandler producer coroutine that sends the SAXInfosetEvents to this class
+ * @param dp DataProcessor that we use to kickstart the unparse process to consumes the events
+ * @param output  OutputChannel where the unparsed data is written
+ * @param resolveRelaiveInfosetBlobURIs if true, elements with type xs:anyURI type and a
+ *   relativeURI are resolved relative to the classpath. This should only be true when used
+ *   via a TDMLRunner or similar testing infrastructure
  */
 class SAXInfosetInputter(
-  unparseContentHandler: DFDL.DaffodilUnparseContentHandler,
+  unparseContentHandler: DaffodilUnparseContentHandler,
   dp: DFDL.DataProcessor,
-  output: DFDL.Output)
-  extends InfosetInputter with DFDL.ConsumerCoroutine {
+  output: DFDL.Output,
+  resolveRelativeInfosetBlobURIs: Boolean)
+  extends InfosetInputter with Coroutine[Array[SAXInfosetEvent]] {
 
   /**
-   * allows support for converting relative URIs in Blobs to absolute URIs, this is necessary
-   * for running TDML tests as they allow relative URIs. Because Daffodil proper only uses
-   * absolute URIs, we hide this functionality behind this flag. It can be set to true by calling
-   * the unparseContentHandler.enableInputterResolutionOfRelativeInfosetBlobURIs(), which calls the
-   * inputter's enableResolutionOfRelativeInfosetBlobURIs() function to set the below variable to true
+   * The index into batchedInfosetEvents that the InfosetInputter is currently returning
+   * information about to the unparse()
    */
-  private var resolveRelativeInfosetBlobURIs: Boolean = false
-
-  private var endDocumentReceived = false
   private var currentIndex: Int = 0
+
+  /**
+   * The array of batched infoset events received from the ContentHandler.
+   */
```

Comment:
Yep, there would be one `resume()` call passing the entire batch to the infoset inputter coroutine, and one `resumeFinal()` call passing the `UnparseResult` back to the ContentHandler. Unfortunately there's still a thread created so there is that same Thread overhead.

In theory, if we found the EndDocument in the first batch we could potentially even avoid the coroutine/thread stuff altogether. Since the SAXInofsetINputter threated wouldn't have been started yet we could have a setter to directly set the batch, and then call the unparse() function passing in that inputter. Could potentially be a big win for small-ish infosets. But does add extra complication.